src/Lights/Shadows/shadowGenerator.ts

```diff
@@ -726,10 +726,16 @@ export class ShadowGenerator implements IShadowGenerator {
             this._shadowMap.renderList = [];
         }
 
-        this._shadowMap.renderList.push(mesh);
-
-        if (includeDescendants) {
-            this._shadowMap.renderList.push(...mesh.getChildMeshes());
+        var renderListMap: Record<string, undefined> = {};
+        this._shadowMap.renderList.forEach(

```

Comment:
I do not feel this is required (and it will save a full loop) as by default values will be undefined

Reply:
This code with spread operators may look tricky, I was trying to use minimal amount of memory (so I used `undefined` instead of `true` as value to save few bytes) to create map of existing meshes containing only keys without values and later look if mesh have been previously added to renderList at constant time. I don't see how checking if mesh is already in list can be performed without couple of traversals for initial mesh and it's children with indexOf() if map would not be created. The only possible optimization I see is not to create map if includeDescendants is false. Please correct me if I misunderstood something in your comments.

Reply:
That's what I ment by not creating map for case when `includeDescendants` is false

```   
        if (includeDescendants) {
            var renderListMap: Record<string, undefined> = {};
            this._shadowMap.renderList.forEach(
                (renderedMesh) => { renderListMap[renderedMesh.name] = undefined; },
            );
            if (!renderListMap.hasOwnProperty(mesh.name)) {
                this._shadowMap.renderList.push(mesh);
            }
            for (var childMesh of mesh.getChildMeshes()) {
                if (!renderListMap.hasOwnProperty(childMesh.name)) {
                    this._shadowMap.renderList.push(childMesh);
                }
            }
        } else {
            var index = this._shadowMap.renderList.indexOf(mesh);
            if (index === -1) {
                this._shadowMap.renderList.push(mesh);
            }
        }
```

Reply:
However I may have overestimated the difference between accessing `object` property and `array.indexOf` for small/medium sized amounts of data, https://stackoverflow.com/a/48031382/8303988, https://jsben.ch/YEecD V8 seem to be doing a lot of optimizations. I'm ready to rewrite it with multiple `indexOf` calls, that way GC wouldn't have any additional work to do.

Reply:
That would be neat :)